alchemist/alchemist-fxui/src/main/kotlin/it/unibo/alchemist/boundary/monitors/SimulationInteractions.kt

```diff
@@ -0,0 +1,753 @@
+import java.awt.Point
+import java.util.Timer
+import java.util.concurrent.Semaphore
+import kotlin.concurrent.fixedRateTimer
+import kotlin.math.abs
+import kotlin.math.max
+import kotlin.math.min
+import kotlin.math.roundToInt
+
+/**
+ * An interaction manager that controls the input/output on the environment done through the GUI.
+ * @param monitor the monitor
+ */
+@SuppressFBWarnings(
+    "SF_SWITCH_NO_DEFAULT",
+    "Because the analysis only looks at the generated bytecode, " +
+        "this warning can be incorrect triggered if the default " +
+        "case is at the end of the switch statement and the switch" +
+        " statement doesn't contain break statements for other cases."
+)
+class InteractionManager<T, P : Position2D<P>>(
+    private val monitor: AbstractFXDisplay<T, P>
+) {
+    /**
+     * Describes a certain interaction that has a feedback associated to it.
+     */
+    private enum class Interaction {
+        HIGHLIGHT_CANDIDATE,
+        HIGHLIGHTED,
+        SELECTION_BOX
+    }
+
+    /**
+     * The current environment.
+     */
+    var environment: Environment<T, P>? = null
+    /**
+     * The nodes in the environment.
+     * Should be updated as frequently as possible to ensure a representation of
+     * the feedback that is consistent with the actual environment.
+     */
+    var nodes: Map<Node<T>, P> = emptyMap()
+    /**
+     * The keyboard listener.
+     */
+    val keyboardListener: KeyboardActionListener
+        get() = keyboard.listener
+
+    private lateinit var wormhole: BidimensionalWormhole<P>
+    private lateinit var zoomManager: ZoomManager
+    private val keyboard: KeyboardEventDispatcher = SimpleKeyboardEventDispatcher()
+    private val keyboardPan: DigitalPan<P> by lazy {
+        DigitalPan(wormhole = wormhole) {
+            monitor.repaint()
+            repaint()
+        }
+    }
+    private val mouse: DynamicMouseEventDispatcher = NodeBoundMouseEventDispatcher(monitor)
+    private lateinit var mousePan: AnalogPan
+    private val highlighter = Canvas()
+    private val selector = Canvas()
+    private val selectionHelper: SelectionHelper<T, P> = SelectionHelper()
+    private val selection: ObservableMap<Node<T>, P> = FXCollections.observableHashMap()
+    private val selectionCandidates: ObservableMap<Node<T>, P> = FXCollections.observableHashMap()
+    private val selectedElements: ImmutableMap<Node<T>, P>
+        get() = ImmutableMap.copyOf(selection)
+    private val selectionCandidatesMutex: Semaphore = Semaphore(1)
+    @Volatile private var feedback: Map<Interaction, List<() -> Unit>> = emptyMap()
+    private val runMutex: Semaphore = Semaphore(1)
+    /**
+     * The canvases used for input/output.
+     */
+    val canvases = Group().apply { listOf(highlighter, selector).forEach { children.add(it) } }
+
+    /**
+     * Invokes a given command on the simulation.
+     */
+    private val invokeOnSimulation: (Simulation<T, P>.() -> Unit) -> Unit
```

Comment:
No command is provided, this is a val: what is this value supposed to do? Explain it a bit better in the comment...